File: gameserver/src/main/java/l2s/gameserver/skills/effects/i_target_me.java
package l2s.gameserver.skills.effects;

import l2s.gameserver.model.Creature;
import l2s.gameserver.model.Skill;
import l2s.gameserver.templates.skill.EffectTemplate;

public class i_target_me extends i_abstract_effect
{
	public i_target_me(Creature creature, Creature target, Skill skill, boolean reflected, EffectTemplate template)
	{
		super(creature, target, skill, reflected, template);
	}

	@Override
	public boolean checkCondition()
	{
		if(getEffected() == getEffector())
			return false;

		return super.checkCondition();
	}

	@Override
	public void instantUse()
	{
		getEffected().setTarget(getEffector());
		getEffected().abortCast(true, true);
		getEffected().abortAttack(true, true);
		getEffected().getAI().clearNextAction();
	}
}

File: gameserver/src/main/java/l2s/gameserver/handler/effects/impl/instant/retail/i_target_me.kt
package l2s.gameserver.handler.effects.impl.instant.retail

import l2s.gameserver.handler.effects.impl.instant.i_abstract_effect
import l2s.gameserver.model.Creature
import l2s.gameserver.model.Skill
import l2s.gameserver.model.actor.instances.player.Cubic
import l2s.gameserver.stats.Formulas
import l2s.gameserver.templates.skill.EffectTemplate
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Target Me Probability effect implementation.
 * @author Adry_85
 */
class i_target_me(template: EffectTemplate) : i_abstract_effect(template) {

    private val chance = params.getDouble("i_target_me_param1")

    override fun calcSuccess(caster: Creature, target: Creature, skill: Skill): Boolean {
        if (!target.isCreature) {
            return false
        }

        if (!Formulas.calcProbability(chance, caster, target.asCreature(), skill)) {
            return false
        }

        return true
    }

    override fun instantUse(caster: Creature, target: Creature, soulShotUsed: AtomicBoolean, reflected: Boolean, cubic: Cubic) {
        val targetPlayable = target.asPlayable() ?: return

        if (targetPlayable.target != caster) {
            targetPlayable.target = caster

            /* need ?
            target.abortCast(true, true)
            target.abortAttack(true, true)

            target.ai.clearNextAction()*/
        }
    }

}